File: bonus-go/orchestration_patterns.go
// Multi-Agent Orchestration Patterns - Go Implementation
// Microsoft Agent Framework Series - Part 7 Equivalent

package main

import (
	"context"
	"fmt"
	"os"
	"regexp"
	"strings"
	"sync"

	"github.com/joho/godotenv"
	openai "github.com/sashabaranov/go-openai"
)

// Agent represents an AI agent
type Agent struct {
	Name         string
	Instructions string
	client       *openai.Client
}

func NewAgent(client *openai.Client, name, instructions string) *Agent {
	return &Agent{
		Name:         name,
		Instructions: instructions,
		client:       client,
	}
}

func (a *Agent) Run(ctx context.Context, input string) (string, error) {
	resp, err := a.client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model: os.Getenv("AZURE_OPENAI_DEPLOYMENT_NAME"),
		Messages: []openai.ChatCompletionMessage{
			{Role: openai.ChatMessageRoleSystem, Content: a.Instructions},
			{Role: openai.ChatMessageRoleUser, Content: input},
		},
	})
	if err != nil {
		return "", err
	}
	return resp.Choices[0].Message.Content, nil
}

// SequentialOrchestrator runs agents in sequence
type SequentialOrchestrator struct {
	Agents []*Agent
}

func (o *SequentialOrchestrator) Run(ctx context.Context, input string) (string, error) {
	result := input
	var err error

	for _, agent := range o.Agents {
		fmt.Printf("  → Running %s...\n", agent.Name)
		result, err = agent.Run(ctx, result)
		if err != nil {
			return "", err
		}
	}

	return result, nil
}

// ConcurrentOrchestrator runs agents in parallel
type ConcurrentOrchestrator struct {
	Agents     []*Agent
	Aggregator func([]string) string
}

func (o *ConcurrentOrchestrator) Run(ctx context.Context, input string) (string, error) {
	fmt.Printf("  → Running %d agents in parallel...\n", len(o.Agents))

	results := make([]string, len(o.Agents))
	var wg sync.WaitGroup
	var mu sync.Mutex
	var firstErr error

	for i, agent := range o.Agents {
		wg.Add(1)
		go func(idx int, a *Agent) {
			defer wg.Done()
			result, err := a.Run(ctx, input)
			mu.Lock()
			if err != nil && firstErr == nil {
				firstErr = err
			}
			results[idx] = result
			mu.Unlock()
		}(i, agent)
	}

	wg.Wait()

	if firstErr != nil {
		return "", firstErr
	}

	return o.Aggregator(results), nil
}

// HandoffOrchestrator routes to different agents based on responses
type HandoffOrchestrator struct {
	InitialAgent   *Agent
	Agents         map[string]*Agent
	HandoffPattern *regexp.Regexp
}

func (o *HandoffOrchestrator) Run(ctx context.Context, input string) (string, error) {
	currentAgent := o.InitialAgent
	result, err := currentAgent.Run(ctx, input)
	if err != nil {
		return "", err
	}

	for {
		match := o.HandoffPattern.FindStringSubmatch(result)
		if match == nil {
			break
		}

		nextAgentName := match[1]
		nextAgent, exists := o.Agents[nextAgentName]
		if !exists {
			fmt.Printf("  ⚠️ Unknown agent: %s\n", nextAgentName)
			break
		}

		fmt.Printf("  → Handoff to %s\n", nextAgentName)
		currentAgent = nextAgent
		result, err = currentAgent.Run(ctx, input)
		if err != nil {
			return "", err
		}
	}

	return result, nil
}

func main() {
	godotenv.Load()

	config := openai.DefaultAzureConfig(
		os.Getenv("AZURE_OPENAI_API_KEY"),
		os.Getenv("AZURE_OPENAI_ENDPOINT"),
	)
	config.APIVersion = "2024-08-01-preview"
	client := openai.NewClientWithConfig(config)
	ctx := context.Background()

	// Sequential Demo
	fmt.Println("\n=== Sequential Orchestration ===")
	sequential := &SequentialOrchestrator{
		Agents: []*Agent{
			NewAgent(client, "Researcher", "Research the topic and provide key facts."),
			NewAgent(client, "Writer", "Write an article from the research."),
			NewAgent(client, "Editor", "Edit and polish the article."),
		},
	}
	seqResult, _ := sequential.Run(ctx, "AI agents in enterprise software")
	fmt.Printf("Result: %s...\n", truncate(seqResult, 200))

	// Concurrent Demo
	fmt.Println("\n=== Concurrent Orchestration ===")
	concurrent := &ConcurrentOrchestrator{
		Agents: []*Agent{
			NewAgent(client, "MarketAnalyst", "Analyze market trends."),
			NewAgent(client, "TechAnalyst", "Analyze technology aspects."),
			NewAgent(client, "FinanceAnalyst", "Analyze financial health."),
		},
		Aggregator: func(results []string) string {
			var combined strings.Builder
			sections := []string{"## Market", "## Technical", "## Financial"}
			for i, r := range results {
				combined.WriteString(fmt.Sprintf("%s\n%s\n\n", sections[i], r))
			}
			return combined.String()
		},
	}
	concResult, _ := concurrent.Run(ctx, "Analyze Microsoft")
	fmt.Printf("Result: %s...\n", truncate(concResult, 200))

	// Handoff Demo
	fmt.Println("\n=== Handoff Orchestration ===")
	handoff := &HandoffOrchestrator{
		InitialAgent: NewAgent(client, "Tier1",
			"Handle basic queries. Say HANDOFF:Tier2 for technical issues."),
		Agents: map[string]*Agent{
			"Tier2": NewAgent(client, "Tier2", "Handle technical issues in detail."),
		},
		HandoffPattern: regexp.MustCompile(`HANDOFF:(\w+)`),
	}
	handoffResult, _ := handoff.Run(ctx, "The API returns 500 errors")
	fmt.Printf("Result: %s\n", handoffResult)
}

func truncate(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}
	return s[:maxLen] + "..."
}

File: bonus-typescript/src/orchestration-patterns.ts
/**
 * Multi-Agent Orchestration Patterns - TypeScript
 * Microsoft Agent Framework Series - Part 7 Equivalent
 */

import OpenAI from 'openai';

type Agent = {
    name: string;
    instructions: string;
    run: (input: string) => Promise<string>;
};

// Create an agent
function createAgent(client: OpenAI, name: string, instructions: string): Agent {
    return {
        name,
        instructions,
        run: async (input: string) => {
            const response = await client.chat.completions.create({
                model: 'gpt-4o',
                messages: [
                    { role: 'system', content: instructions },
                    { role: 'user', content: input }
                ]
            });
            return response.choices[0].message.content || '';
        }
    };
}

// Sequential Orchestrator
class SequentialOrchestrator {
    private agents: Agent[];

    constructor(agents: Agent[]) {
        this.agents = agents;
    }

    async run(input: string): Promise<string> {
        let result = input;

        for (const agent of this.agents) {
            console.log(`  → Running ${agent.name}...`);
            result = await agent.run(result);
        }

        return result;
    }
}

// Concurrent Orchestrator
class ConcurrentOrchestrator {
    private agents: Agent[];
    private aggregator: (results: string[]) => string;

    constructor(agents: Agent[], aggregator: (results: string[]) => string) {
        this.agents = agents;
        this.aggregator = aggregator;
    }

    async run(input: string): Promise<string> {
        console.log(`  → Running ${this.agents.length} agents in parallel...`);

        const promises = this.agents.map(agent => agent.run(input));
        const results = await Promise.all(promises);

        return this.aggregator(results);
    }
}

// Handoff Orchestrator
class HandoffOrchestrator {
    private initialAgent: Agent;
    private agents: Map<string, Agent>;
    private handoffPattern: RegExp;

    constructor(
        initialAgent: Agent,
        agents: Record<string, Agent>,
        handoffPattern: RegExp = /HANDOFF:(\w+)/
    ) {
        this.initialAgent = initialAgent;
        this.agents = new Map(Object.entries(agents));
        this.handoffPattern = handoffPattern;
    }

    async run(input: string): Promise<string> {
        let currentAgent = this.initialAgent;
        let result = await currentAgent.run(input);

        while (true) {
            const match = result.match(this.handoffPattern);
            if (!match) break;

            const nextAgentName = match[1];
            const nextAgent = this.agents.get(nextAgentName);

            if (!nextAgent) {
                console.log(`  ⚠️ Unknown agent: ${nextAgentName}`);
                break;
            }

            console.log(`  → Handoff to ${nextAgentName}`);
            currentAgent = nextAgent;
            result = await currentAgent.run(input);
        }

        return result;
    }
}

// Demo
async function main() {
    const client = new OpenAI({
        apiKey: process.env.AZURE_OPENAI_API_KEY,
        baseURL: `${process.env.AZURE_OPENAI_ENDPOINT}/openai/deployments/gpt-4o`,
        defaultQuery: { 'api-version': '2024-08-01-preview' }
    });

    // Sequential Demo
    console.log('\n=== Sequential Orchestration ===');
    const researcher = createAgent(client, 'Researcher', 'Research the topic and provide key facts.');
    const writer = createAgent(client, 'Writer', 'Write an article from the research.');
    const editor = createAgent(client, 'Editor', 'Edit and polish the article.');

    const sequential = new SequentialOrchestrator([researcher, writer, editor]);
    const seqResult = await sequential.run('AI agents in enterprise software');
    console.log(`Result: ${seqResult.substring(0, 200)}...`);

    // Concurrent Demo
    console.log('\n=== Concurrent Orchestration ===');
    const market = createAgent(client, 'MarketAnalyst', 'Analyze market trends.');
    const tech = createAgent(client, 'TechAnalyst', 'Analyze technology aspects.');
    const finance = createAgent(client, 'FinanceAnalyst', 'Analyze financial health.');

    const concurrent = new ConcurrentOrchestrator(
        [market, tech, finance],
        (results) => results.map((r, i) => `## Section ${i + 1}\n${r}`).join('\n\n')
    );
    const concResult = await concurrent.run('Analyze Microsoft');
    console.log(`Result: ${concResult.substring(0, 200)}...`);

    // Handoff Demo
    console.log('\n=== Handoff Orchestration ===');
    const tier1 = createAgent(client, 'Tier1',
        'Handle basic queries. Say HANDOFF:Tier2 for technical issues.');
    const tier2 = createAgent(client, 'Tier2', 'Handle technical issues.');

    const handoff = new HandoffOrchestrator(tier1, { Tier2: tier2 });
    const handoffResult = await handoff.run('The API returns 500 errors');
    console.log(`Result: ${handoffResult}`);
}

main().catch(console.error);
